src/samples/Frame.hpp

```diff
@@ -12,272 +12,298 @@
-	     */
-	    uint32_t getPixelSize() const;
-	    
-	    /**
-	     * Returns the size of a row (in bytes). This takes into account the image
-	     * mode as well as the data depth.
-	     * @return Number of channels * width * bytes used to represent one colour
+        MODE_PJPG = COMPRESSED_MODES + 1,
+        MODE_JPEG = COMPRESSED_MODES + 2,
+        MODE_PNG = COMPRESSED_MODES + 3
+    };
+
+    enum frame_status_t {
+        STATUS_EMPTY,
+        STATUS_VALID,
+        STATUS_INVALID
+    };
+
+    /* A single image frame */
+    struct Frame
+    {
+    public:
+        /**
+         * Initialize the frame
+         */
+        Frame();
+
+        /**
+         * @param width the frame width
+         * @param height the frame height
+         * @param depth the bit depth
+         * @param mode the pixel mode
+         * @param val a value used to initialize each byte of the image's buffer.
+         *     if negative, it is left uninitialized
+         * @param sizeInBytes the desired size in bytes of the image's buffer. If zero,
+         *     it is computed from the other parameters
+         */
+        Frame(uint16_t width, uint16_t height, uint8_t depth=8, frame_mode_t mode=MODE_GRAYSCALE, uint8_t const val = 0,size_t sizeInBytes=0);
+
+        /**
+         * @param other the frame we copy from
+         * @param bcopy if true, both metadata (size, mode, ...) and data are copied.
+         *     If false, the data itself is not
+         */
+        Frame(const Frame &other, bool bcopy = true);
+
+        //copies all attributes which are independent from size and mode
+        //if an attribute already exists the old value is over written
+        void copyImageIndependantAttributes(const Frame &other);
+
+        /** Copy information and optionally data from another frame
+         *
+         * @param other the frame we copy from
+         * @param bcopy if true, both metadata (size, mode, ...) and data are copied.
+         *     If false, the data itself is not
+         */
+        void init(const Frame &other,bool bcopy = true);
+
+        /** Initialize the image
+         *
+         * @param width the frame width
+         * @param height the frame height
+         * @param depth the bit depth
+         * @param mode the pixel mode
+         * @param val a value used to initialize each byte of the image's buffer.
+         *     if negative, it is left uninitialized
+         * @param sizeInBytes the desired size in bytes of the image's buffer. If zero,
+         *     it is computed from the other parameters
+         */
+        void init(uint16_t width, uint16_t height, uint8_t depth=8, frame_mode_t mode=MODE_GRAYSCALE, int16_t val = 0, size_t sizeInBytes=0);
+
+        /** Fill the image buffer's bytes by a given value
+         *
+         * @param val a value used to initialize each byte of the image's buffer.
+         *     if negative, it is left uninitialized
+         */
+        void reset(int val = 0);
```

Comment:
constness of `val` should be removed in the cpp-file as well.

The current implementation of `reset` assumes 8-bit images (for 16bit images the upper and lower byte will be set equal). It would be possible to generalize this using `std::fill_n` (not sure if this is needed, probably the only actual use-cases are `0` and `0xff`)
Instead of changing the implementation, I'd also be fine with documenting the current behavior)

Reply:
That's of course right ... but then it built on my machine ... hum

Reply:
> Instead of changing the implementation, I'd also be fine with documenting the current behavior)

That's why I wrote "each byte of the image buffer". But I'll be more explicit.

Reply:
Updated

Reply:
Actually, C++ appears to ignore const-ness of POD parameters (except inside the function body of course). And different const-ness in declaration and implementation does not throw warnings (not even with `-Weverything` on clang), so allowing this seems to be intentional (https://godbolt.org/z/bGPPT8hT3).

Reply:
Making a by-value parameter const does not change anything outside of the function itself, and in c++, all parameters are by-value(references and pointers are not special, the address is passed by-value). Another thing that is only relevant to the signature preceding the implementation are parameter names, and as such, you can have different names in forward declarations, or have none at all. Especially different names are confusing to no end.

For consistency and sanity, it is probably good to keep the signatures in synch.